README

0.  Availability
============

The source code for this package is available from
http://research-pub.gene.com/gmap.  License terms are provided in the
COPYING file.


1.  Building and installing GMAP and GSNAP
==========================================

Prerequisites: a Unix system (including Cygwin on Windows), a C
compiler, and Perl

Step 1: Set your site-specific variables by editing the file
config.site.  In particular, you should set appropriate values for
"prefix" and probably for "with_gmapdb", as explained in that file.
If you are compiling this package on a Macintosh, you may need to edit
CFLAGS to be

CFLAGS = '-O3 -m64'

since Macintosh machines will make only 32-bit executables by default.


Step 2: Build, test, and install the programs, by running the
following GNU commands

    ./configure
    make
    make check   (optional)
    make install

Note 1: Instead of editing the config.site file in step 1, you may type
everything on the command line for the ./configure script in step 2,
like this

    ./configure --prefix=/your/usr/local/path --with-gmapdb=/path/to/gmapdb

If you omit --with-gmapdb, it defaults to ${prefix}/share.  If you
omit --prefix, it defaults to /usr/local.  Note that on the command
line, it is "with-gmapdb" with a hyphen, but in a config.site file,
it is "with_gmapdb" with an underscore.


Note 2: If you want to keep your version of config.site or have
multiple versions, you can save the file to a different filename, and
then refer to it like this

    ./configure CONFIG_SITE=<config site file>


Note 3: GSNAP is designed for short reads of a limited length, and
uses a configure variable called MAX_READLENGTH (default 300) as a
guide to the maximum read length.  You may set this variable by
providing it to configure like this

    ./configure MAX_READLENGTH=<length>

or by defining it in your config.site file (or in the file provided to
configure as the value of CONFIG_SITE).  Or you may set the value of
MAX_READLENGTH as an environment variable before calling ./configure.
If you do not set MAX_READLENGTH, it will have the default value shown
when you run "./configure --help".

Note that MAX_READLENGTH applies only to GSNAP.  GMAP, on the other
hand, can process queries up to 1 million bp.

Also, starting with version 2014-08-20, if your C compiler can
handle stack-based memory allocation using the alloca() function,
GSNAP ignores MAX_READLENGTH, and can handle reads longer than that
value.


Note 4: GSNAP can read from gzip-compressed FASTA or FASTQ input
files.  This feature requires the zlib library to be present
(available from http://www.zlib.net).  The configure program will
detect the availability of zlib automatically.  However, to disable
this feature, you can add "--disable-zlib" to the ./configure command
or edit your config.site file to have the command "disable_zlib".


Note 5: GSNAP can read from bzip2-compressed FASTA or FASTQ input
files.  This feature requires the bzlib library to be present.  The
configure program will detect the availability of bzlib automatically.
However, to disable this feature, you can add "--disable-bzlib" to the
./configure command or edit your config.site file to have the command
"disable_bzlib".


Note 6: GSNAP optionally supports the Goby input and output file
formats.  To implement this functionality, you need to obtain and
compile the libraries from http://campagnelab.org/software/goby.  If
the resulting header files are located in /path/to/goby/include and
the library files are in /path/to/goby/lib, you can then add the flag
"--with-goby=/path/to/goby" to your ./configure command or edit your
config.site file to have this directory as the value for "with_goby".


2.  Possible issues during compilation
======================================

Recent versions of GMAP and GSNAP use certain techniques to achieve
increased speed.  One of these techniques is special SIMD
(single-instruction, multiple data) instructions that can perform some
computations in parallel.  There are various levels of SIMD
instructions, and we use those from the SSE2 and SSE4.1 instruction
sets.  Most computers built within the last 10 years should support
SSE2 at least.  However, you may run into the following issues:


Compiler issue 1.  If you compile the code successfully, but when you
run the program, you see something that says "Illegal instruction",
then you must be running GMAP or GSNAP on a different computer than
the one where you compiled it.  You may be using a computer cluster
with a variety of different computer models.  Each time you start GMAP
or GSNAP, the program tests to see if the computer can run the same
set of SSE2 or SSE4.1 instructions as were available when the programs
were compiled.  (The programs also check for if the popcnt
instructions work, but popcnt is so widely implemented that they
generally do not cause any problems.)

In that case, you may need to compile your program for the lowest
common denominator by disabling SSE2 instructions by providing
--disable-sse4.1 or --disable-sse2 to ./configure as necessary.
Alternatively, your computer cluster may have the ability to detect
the capabilities of each computer when it receives a job.  Then, you
may want to create different compiled versions of GMAP and GSNAP, and
call the appropriate binary for that particular job.  You will have to
work with your system administrator if you want to accomplish this.


Compiler issue 2.  The most recent versions of GSNAP (starting with
2013-10-01) build a suffix array to help with speed.  If your computer
does not have sufficient RAM, it is possible that the gmap_build 
step fails after printing something like this:

    SACA_K called with n = 3095693984, K = 5, level 0

If this happens, you either need to find a computer with more RAM, or
you can build a genome without a suffix array, by providing
--no-sarray to gmap_build.  The GSNAP program will still work with the
resulting genome index, but it won't achieve optimal speed.  GMAP does
not use the suffix array at all.  Also, large genomes of over 4
billion bp also will not create or use a suffix array.



3.  Downloading a pre-built GMAP/GSNAP database
===============================================

A GMAP/GSNAP "database" is a set of genomic index files, representing
the genome in a hash table format.  You can use the gmap_build program
to build your own database (as described below), but you can started
quickly by downloading a pre-built GMAP/GSNAP database from the same
place you obtained the GMAP program (see above for URL).

Place the database in the GMAPDB directory you specified in the
config.site file when you built the gmap program.  You should include
a subdirectory for each GMAP database; for example, if you downloaded
a database called <genome>, your directory structure should look like
this

    /path/to/gmapdb/<genome>/
    /path/to/gmapdb/<genome>/<genome>.chromosome
    /path/to/gmapdb/<genome>/<genome>.chromosome.iit
    ...
    /path/to/gmapdb/<genome>/<genome>.version

Note that the GMAP database format has changed with the 2013-10-01
release to add a suffix array and other new file formats.  However,
GMAP and GSNAP are both backwards and forwards compatible, meaning
that new source code will work with older genome indices and that old
source code will work with newer genome indices.  Mixing up the two,
though, will result in slower running speed.  To achieve optimal
speed, you should use both the latest source code and rebuild your
genome indices with the latest gmap_build program.  You can tell if
your database has the most recent format if it has files of the type
<genome>.sarray, <genome>.ref12153bitpackptrs, and
<genome>.ref12153bitpackcomp.

Note that the hg19 database provided on the Web site lacks the suffix
array, although GMAP and GSNAP will work fine, albeit more slowly.
The suffix array for hg19 requires 15 GB of disk space.  If you want
the suffix array, you will have to build it yourself, as discussed in
the next section.



4.  Setting up to build a GMAP/GSNAP database (one chromosome per FASTA entry)
==============================================================================

You can also build your own genomic database, using the gmap_build
program provided with this package.  (Note: the gmap_setup
method from older versions is no longer provided or supported.)

Previous versions limited the total sequence length in your database
to 2^32 = 4,294,967,296 (about 4 billion) bp.  However, starting with
version 2013-04-01, a total "genome" may now contain up to 2^64 bp.
However, each individual chromosome is still limited to 2^32 bp.  The
utility programs below will automatically recognize when a genome is
larger than 2^32 bp and build the index files accordingly.  Below I
use the "genome" and "chromosome", but the input sequences can be
anything you wish to align to, including transcripts or small
fragments.

You will need to start with a set of FASTA files containing either
entire chromosomes or contigs that represent pieces of chromosomes.
For example, for the human genome, you can retrieve all of the FASTA
files under the ftp directory at

    ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/chromosomes/

If your FASTA entries each contain a single chromosome, and the
accession for each chromosome is the chromosome number/letter, you can
simply run this command

    gmap_build -d <genome> [-k <kmer size>] <fasta_files...>

which will build and install the GMAP index files in your default
GMAPDB location.

You can see the full usage of gmap_build by doing "gmap_build --help",
but here are some useful flags.  If your FASTA files are compressed
using gzip, you can add the flag "-g" to gmap_build.  You can control
the k-mer size for the genomic index with the -k flag, which can range
from 12 to 15.  The default value for -k is 15, but this requires your
machine to have 4 GB of RAM to build the indices.  If you do not have
4 GB of RAM, then you will need to reduce the value of -k or find
another machine.  Here are the RAM requirements for building various
indices:

    k-mer of 12: 64 MB
    k-mer of 13: 256 MB
    k-mer of 14: 1 GB
    k-mer of 15: 4 GB

These are the RAM requirements for building indices, but not to run
the GMAP/GSNAP programs once the indices are built, because the
genomic indices are compressed.  For example, the genomic index for a
k-mer of 15 gives a gammaptrs file of 64 MB and an offsetscomp file of
about 350 MB, much smaller than the 4 GB that would otherwise be
required.  Therefore, you may want to build your genomic index on a
computer with sufficient RAM, and distribute that index to be used by
computers with less RAM.

The amount of compression can be controlled using the -b or --basesize
parameter to gmap_build.  By default, the value for k-mer size is 15,
and the value for basesize is 12.  If you select a different value for
k-mer size, then basesize is made by default to be equal to that k-mer
size.

If you want to build your genomic databases with more than one k-mer
size, you can re-run gmap_build with different values of -k.  This
will overwrite only the identical files from the previous runs.  You
can then choose the k-mer size at run-time by using the -k flag for
either GMAP or GSNAP.

Finally, you can provide information to gmap_build that certain
chromosomes are circular, with the -c or --circular flag.  The value
for these flags is a list of chromosomes, separated by commas.  The
gmap_build program will then allow GSNAP and GMAP to align reads
across the ends of the chromosome.  For example, the mitochondrial
genome in human beings is circular.



5.  Setting up to build a GMAP/GSNAP database (more complex cases)
==================================================================

If you are indexing a genome, where each chromosome is contained in a
separate FASTA entry, or a set of contigs, where the contigs are
independent of each other, then you can just skip to section 6.
Otherwise, if you have more complicated needs, you may need to add
options to gmap_build.

Note that the term

    <fasta_file>... 

above indicates that multiple files can be listed.  The files can be
in any order, and the contigs can be in any order within these files.
By default, the gmap_build process will sort the contigs and
chromosomes into their appropriate "chrom" order.  For the human
genome, this order is 1, 2, ..., 10, 11, ..., 22, X, Y, M, followed by
all other chromosomes in numeric/alphabetical order.  If you don't
want this sort, provide the "-s none" flag to gmap_build.  Other sort
options besides "none" and "chrom" are "alpha" and "numeric-alpha".

You can type "gmap_setup --help" to see the full set of options.  We
discuss some specific situations below.


5a.  Contigs are mapped to chromosomes
======================================

If your FASTA entries consist of contigs, each of which has a mapping
to a chromosomal region in the header, you may add the -C (or
--contigs-are-mapped) flag to gmap_build, like this

    gmap_build -d <genome> -C <fasta_file>...

Then gmap_build will try to parse a chromosomal region from each
header.  The program knows how to parse the following patterns:

    chr=1:217281..257582  [may insert spaces around '=', or omit '=' character]
    chr=1                 [may insert spaces around '=', or omit '=' character]
    chromosome 1                                                         [NCBI .mfa format]
    chromosome:NCBI35:22:1:49554710:1                                    [Ensembl format]
    /chromosome=2                                                        [Celera format]
    /chromosome=2 /alignment=(88840247-88864134) /orientation=rev        [Celera format]
    chr1:217281..257582
    chr1                  [may insert spaces after 'chr']

If only the chromosome is specified, without coordinates, the program
will assign its own chromosomal coordinates by concatenating the
contigs within each chromosome.  If gmap_build cannot figure out the
chromosome, it will assign it to chromosome "NA".


5b.  Using an MD file
=====================

Another possibility is that your FASTA entries consist of contigs that
mapped to chromosomes, where the mapping information is in an external
file.  Genomes from NCBI typically include an ".md" file (like
seq_contig.md) that specifies the chromosomal coordinates for each
contig.  To use this information, provide the -M (or --mdfile) flag to
gmap_build like this

    gmap_build -d <genome> -M <mdfile> <fasta_file>...

The program will then try to parse the mdfile (which often changes
formats) and verify with you which columns contain the contig names
and chromosomal coordinates.


5c.  Compressed FASTA files or files requiring processing
=========================================================

If your genome files are compressed using gzip, you can simply add the
flag "-g" to gmap_build.  But if your genome files require some other
type of processing, you may need to write a small script that pipes
the sequences in FASTA format to gmap_build.  You can tell gmap_build
about your script with the -E (or --fasta-pipe) flag, like this

    gmap_build -d <genome> -E 'gunzip -c chr*.gz'
    gmap_build -d <genome> -E 'cat *.fa | ./add-chromosomal-info.pl'

You can think of the command as a Unix pipe for processing each FASTA
file before it is read by gmap_build.



6.  Running GMAP
================

To see the full set of options, type "gmap --help".  The following are
some common examples of usage.  For more examples, see the document
available at http://www.gene.com/share/gmap/paper/demo-slides.pdf

For each of the examples below, we assume that you have installed a
genome database called <genome> in your GMAPDB directory.  (If your
database is located elsewhere, you can specify the -D flag to gmap or
set the environment variable GMAPDB to point to that directory.)

* Mapping only: To map one or more cDNAs in a FASTA file onto a
  genome, run GMAP as follows:

    gmap -d <genome> <cdna_file>

* Mapping and alignment: If you want to map the cDNAs to a genome,
  and show the full alignment, provide the -A flag:

    gmap -d <genome> -A <cdna_file>


* Alignment only: To align one or more cDNAs in a FASTA file onto a
  given genomic segment (also in a FASTA file), use the -g flag
  instead of the -d flag:

    gmap -g <genome_segment> -A <cdna_file>


* Batch mode: If you have a large number of cDNAs to run, and you have
  sufficient RAM to run in batch mode, add the "-B 3", "-B 4", or "-B
  5" option.  Details for these options are provided by running "gmap
  --help".

    gmap -d <genome> -B 5 -A <cdna_file>


* Multithreaded mode: If your machine has several processors, you can
  make batch mode run even faster by specifying multiple threads with
  the -t flag:

    gmap -d <genome> -B 5 -A -t <nthreads> <cdna_file>

  Note that with multiple threads, the output results will appear in
  random order, depending on which thread finishes its computation
  first.  If you wish your output to be in the same order as the 
  input cDNA file, add the '-O' (letter O, not the number 0) flag
  to get ordered output.

  Guidelines: The -t flag specifies the number of computational
  threads.  In addition, if your machine supports threads, GMAP also
  uses one thread for reading the input query sequences, and one
  thread for printing the output results.  Therefore, the total number
  of threads will be 2 plus the number you specify.  The program will
  work optimally if it uses one thread per available processor.  If
  you specify too many threads, you can cause your computer to thrash
  and slow down.  Note that other programs running on your computer
  also need processors.


* Compressed output: If you want to store the alignment results in a
  compressed format, use the -Z flag.  You can uncompress the results
  by using the gmap_uncompress.pl program:

    gmap -d <genome> -Z <cdna_file> > x
    cat x | gmap_uncompress


Note that gmap is written for small genomes (less than 2^32 bp in
total length).  With large genomes, there is an equivalent program
called gmapl, which you should run instead of gmap.  The gmapl program
is equivalent to gmap, and is based on the same source code, but is
compiled to use 64-bit index files instead of 32-bit files.  The gmap
and gmapl programs will detect whether the genomes are the correct
size, and will exit if you try to run them on the wrong-sized genomes.



7.  Building map files
======================

This package includes an implementation of interval index trees
(IITs), which permits efficient lookup of interval information.  The
gmap program also allows you (with its -m flag) to look up pre-mapped
annotation information that overlaps your query cDNA sequence.  These
interval index trees (or map files) are built using the iit_store
program included in this package.  To build a map file, do the
following:

Step 1: Put your map information for a given genome into a map file
with the following FASTA-like format:
   
    >label coords optional_tag
    optional_annotation (which may be zero, one, or multiple lines)

For example, the label may be an EST accession, with the coords
representing its genomic position.  Labels may be duplicated if
necessary.

The coords should be of the form

    chr:position
    chr:startposition..endposition

The term "chr:position" is equivalent to "chr:position..position".  If
you want to indicate that the interval is on the minus strand or
reverse direction, then <endposition> may be less than
<startposition>.

Tags are very general and can be used for a variety of purposes.  For
example, you could 


Step 2: Run iit_store on this map file as follows

    cat <mapfile> | iit_store -o <mapname>

The program will create a file called <mapname>.iit.

Now you can retrieve this information with iit_get

    iit_get <mapname>.iit <coords>

where <coords> has the format "chr:position" or
"chr:startposition..endposition".  The iit_get program has other
capabilities, including the ability to retrieve information by label,
like this:

    iit_get <mapname>.iit <label>

More details can be found by running "iit_get --help".

If you plan to use this map information frequently, you may want to
place it with its corresponding genome for future use.  In each

ChangeLog
2015-09-30  twu

    * VERSION: Updated version number

    * stage1hr.c: In converting segments to GMAP, ensuring that the pairs are
      monotonic in genomepos

    * inbuffer.c: Initializing variables

2015-09-28  twu

    * oligoindex_hr.c, public-2015-07-23, src: Merged revision 175543 from trunk
      to handle the case where left_plus_length < indexsize

    * VERSION: Updated version number

    * stage1hr.c: Putting debugging statement inside debug macro

2015-09-22  twu

    * dynprog_genome.c, pair.c, pair.h, public-2015-07-23, src, stage1hr.c,
      stage3.c, stage3hr.c: Merged revisions 174475 through 174480 from trunk to
      improve alignments involving microexons, splices at ends, bridging intron
      gaps, and multiple segments

    * VERSION, public-2015-07-23, src, stage1hr.c, stage3hr.c: Merged revision
      174117 from trunk to fix calls to Genome_get_segment_blocks_left

2015-09-12  twu

    * splice.c: Fixed FREEA statements

    * gmap_build.pl.in, util: Merged revision 173888 from trunk to allow for
      spaces in destination directory

    * gsnap.c, src, stage1hr.c, stage1hr.h, stage3hr.c, stage3hr.h, substring.c,
      substring.h, uniqscan.c: Merged revisions 173889 through 173894 from trunk
      to add --min-coverage and remove --terminal-threshold and
      reject_trimlength.  Changed criteria for running find_terminals.

2015-09-01  twu

    * VERSION, chimera.c, gmap.c, public-2015-07-23, src: Merged revisions
      173188 and 173189 from trunk to set dinucleotides for out-of-bound
      chimeric breakpoints and to set some uninitialized variables

    * stage3hr.c: Merged revision 173165 from trunk to favor non-zero sensedirs
      when sorting results

    * splice.c: Merged revision 173164 from trunk to fix variable names for
      FREEA

    * oligoindex_hr.c: Merged revision 173163 from trunk to initialize some
      return variables when exiting trimming procedure early

    * chimera.c, chimera.h, gmap.c, pair.c: Merged revision 173162 from trunk to
      fix issues when chimeras extend to beginning or end of chromosomes

    * dynprog_genome.c: Merged revision 173161 from trunk to fix bridging intron
      gaps when no probabilities are found

2015-08-31  twu

    * gmap.c, outbuffer.c, src, stage1.c, stage3.c: Merged revision 173034 to
      not use alloca for array of Batch_T objects

2015-08-27  twu

    * stage3.c: Applied revision 172740 from trunk to change criterion for
      evaluating splice neighborhood to allow for short ends

    * sarray-read.c: Applied revision 172478 from trunk to use
      max_mismatches_allowed from original call to suffix array algorithm, and
      not allowing it to be unlimited

    * stage3hr.c: Applied part of revision 172472 from trunk to call
      Genome_get_segment_blocks_left with chroffset and not chrhigh

    * genome.c: Applied revision 172476 from trunk to set end of genomealt
      string to be NULL

    * dynprog_genome.c: Applied revision 172475 from trunk to change loop end
      condition to avoid accessing uninitialized variables

    * gsnap.c, samprint.c, samprint.h: Applied revisions 171145 and 171803 from
      trunk to add flags --add-paired-nomappers and
      --paired-flag-means-concordant

    * oligoindex_hr.c: Applied revision 170792 from trunk to restore missing
      line in counting of 9-mers

    * public-2015-07-23: Created release branch for 2015-07-23

2015-07-23  twu

    * VERSION: Updated version number

    * stage1hr.c: Removed an abort command from debugging

    * sarray-read.c: Using new interface to Bytecoding lcp_next function. 
      Commented out code that is not used when SUBDIVIDE_ENDS is not defined.

    * bytecoding.c, bytecoding.h: Call to lcp_next now returns child_next

    * VERSION: Updated version number

    * dynprog_genome.c: Fixed boundaries that led to negative coordinates for
      splice site candidates.

    * stage1hr.c: Removed unused variables

    * stage1hr.c: Removed allvalidp as parameter to align_end and align_pair.

2015-07-22  twu

    * stage1hr.c: Setting spanningsetp and completesetp to false if querylength
      < min_kmer_readlength

    * stage1hr.c: Removed restriction on min_readlength.  Running only suffix
      array, if possible, if reads are too short.

    * access.c: Changed user message

    * sarray-write.c: Changing plcp[n] to be 0 instead of -1

    * sarray-read.c: Improved debugging results

    * access.c: Printing user message if shmem fails

2015-07-17  twu

    * get-genome.c, sequence.c, sequence.h: Added flags for --stream-chars and
      --stream-ints

2015-06-26  twu

    * 2015-statgen, Makefile.gsnaptoo.am, algorithm.tex, discussion.tex,
      features.tex, introduction.tex, trunk, util: Modified mergeinfo

    * config.site.rescomp.tst: Updated version

    * index.html: Updated for version 2015-06-23

    * archive.html: Updated for version 2014-12-31

    * README: Removed references to Goby

    * access.c, bigendian.c, bigendian.h, bitpack64-access.c, bitpack64-read.c,
      bitpack64-readtwo.c, bytecoding.c, compress.c, compress.h, genome-write.c,
      genome.c, genome.h, genome128_hr.c, iit-read-univ.c, indexdb.c,
      indexdb_hr.c, sarray-read.c, sarray-write.c, snpindex.c, src, types.h,
      univinterval.h: Merged revisions 167282 through 168383 from
      branches/2015-06-10-bigendian to support bigendian architectures

    * Makefile.dna.am, Makefile.util.am: Added instructions for check-bigendian

2015-06-24  twu

    * VERSION, config.site.rescomp.tst: Updated version number

    * algorithm.tex, biblio.bib, discussion.tex, features.tex, introduction.tex,
      toplevel.tex: Final version

    * stage1hr.c: Added comments

    * gmap.c: Removed message about different batch levels

    * gsnap.c: Added option --master-is-worker for MPI version

    * access.c: Using malloc whenever shmget fails

2015-06-15  twu

    * stage1hr.c: Removed extra #endif statements


src/md5.c
static char rcsid[] = "$Id: md5.c 40271 2011-05-28 02:29:18Z twu $";
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif
#ifndef HAVE_MEMCPY
# define memcpy(d,s,n) bcopy((s),(d),(n))
#endif
#ifndef HAVE_MEMMOVE
# define memmove(d,s,n) bcopy((s),(d),(n))
#endif

#include "md5.h"
#include <stdlib.h>
#include <string.h>		/* For memcpy */
#include "mem.h"
#include "types.h"


/* Constants for MD5Transform routine. */
#define S11 7
#define S12 12
#define S13 17
#define S14 22
#define S21 5
#define S22 9
#define S23 14
#define S24 20
#define S31 4
#define S32 11
#define S33 16
#define S34 23
#define S41 6
#define S42 10
#define S43 15
#define S44 21

static unsigned char PADDING[64] = {
  0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

/* F, G, H and I are basic MD5 functions. */
#define F(x, y, z) (((x) & (y)) | ((~x) & (z)))
#define G(x, y, z) (((x) & (z)) | ((y) & (~z)))
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define I(x, y, z) ((y) ^ ((x) | (~z)))

/* ROTATE_LEFT rotates x left n bits. */
#define ROTATE_LEFT(x, n) (((x) << (n)) | ((x) >> (32-(n))))

/* FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4.
Rotation is separate from addition to prevent recomputation. */
#define FF(a, b, c, d, x, s, ac) { \
 (a) += F ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define GG(a, b, c, d, x, s, ac) { \
 (a) += G ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define HH(a, b, c, d, x, s, ac) { \
 (a) += H ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define II(a, b, c, d, x, s, ac) { \
 (a) += I ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }

/* Encodes input (UINT4) into output (unsigned char). Assumes len is
  a multiple of 4. */
static void 
encode (unsigned char *output, UINT4 *input, unsigned int len) {
  unsigned int i, j;

  for (i = 0, j = 0; j < len; i++, j += 4) {
    output[j] = (unsigned char)(input[i] & 0xff);
    output[j+1] = (unsigned char)((input[i] >> 8) & 0xff);
    output[j+2] = (unsigned char)((input[i] >> 16) & 0xff);
    output[j+3] = (unsigned char)((input[i] >> 24) & 0xff);
  }
  return;
}

/* Decodes input (unsigned char) into output (UINT4). Assumes len is
  a multiple of 4. */
static void 
decode (UINT4 *output, unsigned char *input, unsigned int len) {
  unsigned int i, j;

  for (i = 0, j = 0; j < len; i++, j += 4)
    output[i] = ((UINT4)input[j]) | (((UINT4)input[j+1]) << 8) |
      (((UINT4)input[j+2]) << 16) | (((UINT4)input[j+3]) << 24);
  return;
}

/* MD5 basic transformation. Transforms state based on block. */
static void 
transform (UINT4 state[4], unsigned char block[64]) {
  UINT4 a = state[0], b = state[1], c = state[2], d = state[3], x[16];

  decode(x,block,64);

  /* Round 1 */
  FF (a, b, c, d, x[ 0], S11, 0xd76aa478); /* 1 */
  FF (d, a, b, c, x[ 1], S12, 0xe8c7b756); /* 2 */
  FF (c, d, a, b, x[ 2], S13, 0x242070db); /* 3 */
  FF (b, c, d, a, x[ 3], S14, 0xc1bdceee); /* 4 */
  FF (a, b, c, d, x[ 4], S11, 0xf57c0faf); /* 5 */
  FF (d, a, b, c, x[ 5], S12, 0x4787c62a); /* 6 */
  FF (c, d, a, b, x[ 6], S13, 0xa8304613); /* 7 */
  FF (b, c, d, a, x[ 7], S14, 0xfd469501); /* 8 */
  FF (a, b, c, d, x[ 8], S11, 0x698098d8); /* 9 */
  FF (d, a, b, c, x[ 9], S12, 0x8b44f7af); /* 10 */
  FF (c, d, a, b, x[10], S13, 0xffff5bb1); /* 11 */
  FF (b, c, d, a, x[11], S14, 0x895cd7be); /* 12 */
  FF (a, b, c, d, x[12], S11, 0x6b901122); /* 13 */
  FF (d, a, b, c, x[13], S12, 0xfd987193); /* 14 */
  FF (c, d, a, b, x[14], S13, 0xa679438e); /* 15 */
  FF (b, c, d, a, x[15], S14, 0x49b40821); /* 16 */

 /* Round 2 */
  GG (a, b, c, d, x[ 1], S21, 0xf61e2562); /* 17 */
  GG (d, a, b, c, x[ 6], S22, 0xc040b340); /* 18 */
  GG (c, d, a, b, x[11], S23, 0x265e5a51); /* 19 */
  GG (b, c, d, a, x[ 0], S24, 0xe9b6c7aa); /* 20 */
  GG (a, b, c, d, x[ 5], S21, 0xd62f105d); /* 21 */
  GG (d, a, b, c, x[10], S22,  0x2441453); /* 22 */
  GG (c, d, a, b, x[15], S23, 0xd8a1e681); /* 23 */
  GG (b, c, d, a, x[ 4], S24, 0xe7d3fbc8); /* 24 */
  GG (a, b, c, d, x[ 9], S21, 0x21e1cde6); /* 25 */
  GG (d, a, b, c, x[14], S22, 0xc33707d6); /* 26 */
  GG (c, d, a, b, x[ 3], S23, 0xf4d50d87); /* 27 */
  GG (b, c, d, a, x[ 8], S24, 0x455a14ed); /* 28 */
  GG (a, b, c, d, x[13], S21, 0xa9e3e905); /* 29 */
  GG (d, a, b, c, x[ 2], S22, 0xfcefa3f8); /* 30 */
  GG (c, d, a, b, x[ 7], S23, 0x676f02d9); /* 31 */
  GG (b, c, d, a, x[12], S24, 0x8d2a4c8a); /* 32 */

  /* Round 3 */
  HH (a, b, c, d, x[ 5], S31, 0xfffa3942); /* 33 */
  HH (d, a, b, c, x[ 8], S32, 0x8771f681); /* 34 */
  HH (c, d, a, b, x[11], S33, 0x6d9d6122); /* 35 */
  HH (b, c, d, a, x[14], S34, 0xfde5380c); /* 36 */
  HH (a, b, c, d, x[ 1], S31, 0xa4beea44); /* 37 */
  HH (d, a, b, c, x[ 4], S32, 0x4bdecfa9); /* 38 */
  HH (c, d, a, b, x[ 7], S33, 0xf6bb4b60); /* 39 */
  HH (b, c, d, a, x[10], S34, 0xbebfbc70); /* 40 */
  HH (a, b, c, d, x[13], S31, 0x289b7ec6); /* 41 */
  HH (d, a, b, c, x[ 0], S32, 0xeaa127fa); /* 42 */
  HH (c, d, a, b, x[ 3], S33, 0xd4ef3085); /* 43 */
  HH (b, c, d, a, x[ 6], S34,  0x4881d05); /* 44 */
  HH (a, b, c, d, x[ 9], S31, 0xd9d4d039); /* 45 */
  HH (d, a, b, c, x[12], S32, 0xe6db99e5); /* 46 */
  HH (c, d, a, b, x[15], S33, 0x1fa27cf8); /* 47 */
  HH (b, c, d, a, x[ 2], S34, 0xc4ac5665); /* 48 */

  /* Round 4 */
  II (a, b, c, d, x[ 0], S41, 0xf4292244); /* 49 */
  II (d, a, b, c, x[ 7], S42, 0x432aff97); /* 50 */
  II (c, d, a, b, x[14], S43, 0xab9423a7); /* 51 */
  II (b, c, d, a, x[ 5], S44, 0xfc93a039); /* 52 */
  II (a, b, c, d, x[12], S41, 0x655b59c3); /* 53 */
  II (d, a, b, c, x[ 3], S42, 0x8f0ccc92); /* 54 */
  II (c, d, a, b, x[10], S43, 0xffeff47d); /* 55 */
  II (b, c, d, a, x[ 1], S44, 0x85845dd1); /* 56 */
  II (a, b, c, d, x[ 8], S41, 0x6fa87e4f); /* 57 */
  II (d, a, b, c, x[15], S42, 0xfe2ce6e0); /* 58 */
  II (c, d, a, b, x[ 6], S43, 0xa3014314); /* 59 */
  II (b, c, d, a, x[13], S44, 0x4e0811a1); /* 60 */
  II (a, b, c, d, x[ 4], S41, 0xf7537e82); /* 61 */
  II (d, a, b, c, x[11], S42, 0xbd3af235); /* 62 */
  II (c, d, a, b, x[ 2], S43, 0x2ad7d2bb); /* 63 */
  II (b, c, d, a, x[ 9], S44, 0xeb86d391); /* 64 */

  state[0] += a;
  state[1] += b;
  state[2] += c;
  state[3] += d;

  return;
}


static void
update (UINT4 count[], UINT4 state[], unsigned char *input, unsigned int input_len) {
  unsigned int i, index, part_len;
  static unsigned char buffer[64];

  /* Compute number of bytes mod 64 */
  index = (unsigned int) ((count[0] >> 3) & 0x3F);

  /* Update number of bits */
  if ((count[0] += ((UINT4) input_len << 3)) < ((UINT4) input_len << 3)) {
    count[1]++;
  }
  count[1] += ((UINT4) input_len >> 29);

  part_len = 64 - index;
  
  /* Transform as many times as possible */
  if (input_len >= part_len) {
    memcpy((void *) &(buffer[index]),input,part_len);
    transform(state,buffer);

    for (i = part_len; i + 63 < input_len; i += 64) {
      transform(state,&(input[i]));
    }
    index = 0;

  } else {
    i = 0;
  }

  /* Buffer remaining input */
  memcpy((void *) &(buffer[index]),&(input[i]),input_len-i);
  return;
}


static unsigned char *
finish (UINT4 count[], UINT4 state[]) {
  unsigned char *digest;
  unsigned char bits[8];
  unsigned int index, pad_len;

  /* Save number of bits */
  encode(bits,count,8);

  /* Pad out to 56 mod 64 */
  index = (unsigned int) ((count[0] >> 3) & 0x3f);
  pad_len = (index < 56) ? (56 - index) : (120 - index);
  update(count,state,PADDING,pad_len);

  /* Append length (before padding) */
  update(count,state,bits,8);

  /* Store state in digest */
  digest = (unsigned char *) CALLOC(16,sizeof(unsigned char));
  encode(digest,state,16);

  return digest;
}  


unsigned char *
MD5_compute (unsigned char *input, int input_len) {
  unsigned char *p;
  UINT4 state[4];
  UINT4 count[2];
  int i;

  count[0] = count[1] = 0;
  state[0] = 0x67452301;
  state[1] = 0xefcdab89;
  state[2] = 0x98badcfe;
  state[3] = 0x10325476;

  p = input;
  for (i = 0; i + 16 < input_len; i += 16) {
    update(count,state,p,16);
    p += 16;
  }
  if (i < input_len) {
    update(count,state,p,input_len-i);
  }
  
  return finish(count,state);
}


void
MD5_print (FILE *fp, unsigned char *digest) {
  int i;

  for (i = 0; i < 16; i++) {
    fprintf(fp,"%02x", digest[i]);
  }
  return;
}

